src/GraphQL/Validation/DocumentValidator.ParseArgumentVisitor.cs

```diff
@@ -0,0 +1,135 @@
+using GraphQL.Execution;
+using GraphQL.Types;
+using GraphQLParser.AST;
+using GraphQLParser.Visitors;
+
+namespace GraphQL.Validation;
+
+public partial class DocumentValidator
+{
+    private class ParseArgumentVisitor : ASTVisitor<ParseArgumentVisitor.Context>
+    {
+        public static readonly ParseArgumentVisitor Instance = new();
+
+        private ParseArgumentVisitor() { }
+
+        protected override async ValueTask VisitOperationDefinitionAsync(GraphQLOperationDefinition operationDefinition, Context context)
+        {
+            var type = operationDefinition.Operation switch
+            {
+                OperationType.Query => context.Schema.Query,
+                OperationType.Mutation => context.Schema.Mutation,
+                OperationType.Subscription => context.Schema.Subscription,
+                _ => null
+            };
+            if (type == null)
+                return;
+            context.Types.Push(type);
+            await base.VisitOperationDefinitionAsync(operationDefinition, context).ConfigureAwait(false);
+            context.Types.Pop();
+        }
+
+        protected override async ValueTask VisitInlineFragmentAsync(GraphQLInlineFragment inlineFragment, Context context)
+        {
+            var typeCondition = inlineFragment.TypeCondition;
+            if (typeCondition == null)
+            {
+                await base.VisitInlineFragmentAsync(inlineFragment, context).ConfigureAwait(false);
+            }
+            else
+            {
+                if (context.Schema.AllTypes[typeCondition.Type.Name.Value]?.GetNamedType() is IComplexGraphType type)
+                {
+                    context.Types.Push(type);
+                    await base.VisitInlineFragmentAsync(inlineFragment, context).ConfigureAwait(false);
+                    context.Types.Pop();
+                }
+            }
+        }
+
+        protected override async ValueTask VisitFragmentDefinitionAsync(GraphQLFragmentDefinition fragmentDefinition, Context context)
+        {
+            if (context.Schema.AllTypes[fragmentDefinition.TypeCondition.Type.Name.Value]?.GetNamedType() is IComplexGraphType type)
```

Comment:
🤔 Why all these `is IComplexGraphType` checks? Scalar fields **can** have arguments as well.

Reply:
This is a visitor for a fragment definition.  [5.5.1.3](https://spec.graphql.org/June2018/#sec-Fragments-On-Composite-Types) says:

> The target type of fragment must have kind UNION, INTERFACE, or OBJECT.

We already have error checking for 5.5.1.3 (see `FragmentsOnCompositeTypes`), which will append the proper error to the response.  So here, if 5.5.1.3 was violated we just skip the fragment altogether -- but assuming it was not, it will always cast to an `IComplexGraphType`.

Please add a comment if you wish me to explain another piece of code.

Reply:
OK. I'm talking about all `is IComplexGraphType` checks in this class. I just chose the first check that came to hand.